HACKING.jst.md

```diff
@@ -0,0 +1,100 @@
+  `vendor/diff-parsers-upstream-std.patch`. The diff is used to monitor the
+  drift between the standard parser and the upstream parser; the patch file
+  has been mangled (it is produced by `tools/diff-ocaml` which uses `sed` to
+  alter the output of `diff`) and thus cannot be applied by any known tool.
+  
+    You will not have to interact with `parser-upstream`.
+  
+* `parser-standard` is meant to be very similar to the parser used in the
+  OCaml compiler. Its role in `ocamlformat` is (only) to provide a safety-check:
+  we want to make sure that the formatting does not change the AST as parsed by
+  the compiler. So after formatting, the parsed AST is checked against the
+  original parsed AST to make sure they are the same (modulo some normalization,
+  as written in `lib/Normalize_std_ast.ml`). Key point: this parser and its
+  representation are *not* pretty-printed and never interact with the
+  pretty-printer. The *only* reason this is here is to mimic the behavior of the
+  OCaml compiler. Accordingly, it should be as close to the OCaml compiler's
+  parser as possible.
+  
+* `parser-extended` uses an extended parsetree, capable of storing extra
+  information useful in preserving the structure of the user's code. It is this
+  extended parsetree that gets pretty-printed. The parser here forms part of the
+  core implementation of `ocamlformat`.
+  
+* `parser-recovery` is used for partial parsing, so that a bogus input source
+  file does not get mangled. It was an experiment that has been discontinued by
+  upstream and is not used with Jane Street. It uses the same parsetree as
+  `parser-extended`. A patchfile tracking the changes between `parser-extended`
+  and `parser-recovery` is generated. Just accept any changes that accrue there.
+  
+The directory `vendor/ocaml-common` contains files that are shared between
+`parser-standard` and `parser-extended`, like `location.ml`.
+  
+Testing is done in the `test` directory. TODO: Write more about the testing
+setup.
+
+Design considerations
+---------------------
+
+Because the value of `parser-standard` is entirely in its ability to mimic the
+compiler's parser, we want to keep this parser as close as possible to the
+compiler's. We will want to copy over any changes made to the compiler's parser
+into this version of `ocamlformat`'s parser.
+
+On the other hand, the `parser-extended` can go off in its own direction: its
+parsetree should be designed to make pretty-printing easy. In addition, we want
+to make sure that incorporating upstream changes is easy. We thus feel free to
+edit the parsetree in `parser-extended`, but we do so by adding new
+constructors, not modifying existing ones. In addition, new code should be
+marked off with comments like `(* Jane Street extension *)` and `(* End Jane
+Street extension *)`. Because of the ability to extend the parsetree here, we do
+*not* use jane-syntax in `parser-extended`.
```

Comment:
Perhaps a thing to mention here is the subterm check.  Something like:

`ocamlformat` routinely checks, at various places in `Fmt_ast` that the thing it's about to print is an exact subterm of a "context" we're within (some larger term that we're in the middle of printing).  Make sure you've designed your additions to `parser-extended` so that, while printing, you'll always recurse on exact structural subterms.  If you, for example, print an attribute specially and then try to recurse on a term without that attribute, you're in for a bad time.